### PR TITLE
Lifecycle returns phase specific exit codes

### LINKED ISSUE #350: Update all error codes to match platform/0.4 spec
https://github.com/buildpacks/spec/pull/115

### PR DESCRIPTION
**Pull request by jabrown85**
All the phases that returned exit codes before now return exit codes that match the updated spec.

Closes #350

### CODE REVIEW

File: cmd/exit.go [hunk 1]
```diff
@@ -14,10 +14,14 @@ const (
 	// 5: CodeNotFound
 	CodeFailedDetect           = 100
 	CodeFailedDetectWithErrors = 101
-	CodeFailedBuild            = 7
+	CodeDetectError            = 102
```
Comment (jabrown85) [inline cmd/exit.go:17]: @ekcasey I wasn't sure if we wanted to document this in the spec explicitly or not. If we did, I guess we would want to do that for the other generic phase failures.

File: cmd/exit.go [hunk 2]
```diff
@@ -14,10 +14,14 @@ const (
...
+	CodeAnalyzeError           = 200
+	CodeRestoreError           = 300
```
Comment (ekcasey) [inline cmd/exit.go:19]:
```suggestion
	CodeAnalyzeError           = 202
	CodeRestoreError           = 302
```
Maybe this is super OCD but given that the generic detect error is `102` I like using `x02` for the other phases.

File: cmd/exit.go [hunk 3]
```diff
@@ -14,10 +14,14 @@ const (
...
+	CodeFailedBuild            = 400
 	CodeFailedLaunch           = 8
```
Comment (ekcasey) [inline cmd/exit.go:21]:
```suggestion
	CodeFailedBuild            = 400
	CodeLaunchError            = 702
```

File: cmd/exit.go [hunk 4]
```diff
@@ -14,10 +14,14 @@ const (
...
 	// 9: CodeFailedUpdate
-	CodeFailedSave   = 10
+	CodeFailedExport = 500
+	CodeFailedRebase = 600
```
Comment (ekcasey) [inline cmd/exit.go:24]:
```suggestion
	CodeExportError = 502
	CodeRebaseError = 602
```

File: cmd/exit.go [hunk 5]
```diff
@@ -14,10 +14,14 @@ const (
 	// 5: CodeNotFound
 	CodeFailedDetect           = 100
 	CodeFailedDetectWithErrors = 101
-	CodeFailedBuild            = 7
+	CodeDetectError            = 102
+	CodeAnalyzeError           = 200
+	CodeRestoreError           = 300
+	CodeFailedBuild            = 400
```
Comment (ekcasey) [inline cmd/exit.go:20]: Here I think we need a generic `CodeBuildError` and then different `CodeFailedBuildWithError` (or something like that), set to `401` for buildpack error:

https://github.com/buildpacks/spec/blob/platform/0.4/platform.md#outputs-3

File: cmd/exit.go (hunk 1 above)
Comment (ekcasey) [reply cmd/exit.go:17]: I think it is sufficiently documented to the extent that it is in the detect-specific lifecycle error range (102-199).

File: detector.go [hunk 6]
```diff
@@ -320,23 +320,27 @@ func (bo BuildpackOrder) detect(done, next []Buildpack, optional bool, wg *sync.
...
-			continue
+		switch err := err.(type) {
+		case *Error:
+			if err.Type == ErrTypeBuildpack {
+				buildpackErr = true
+			}
+			if err.Type == ErrTypeBuildpack || err.Type == ErrTypeFailedDetection {
+				wg = &sync.WaitGroup{}
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (ekcasey) [reply detector.go]: `(bg BuildpackGroup) Detect(c *DetectConfig) (BuildpackGroup, BuildPlan, error)` is also exported so we should probably do the error translation there as well. It isn't used anywhere in the lifecycle but it is a potentially useful library method.

**Comment by ekcasey** [comment]
Rebase and merged here: https://github.com/buildpacks/lifecycle/pull/361

**Pull request closed without merging**